MPI Benchmark

franzpoeschel opened the pull request:
Adds a generic MPI benchmark to openPMD (#37).

Todo:

- [x] The reading benchmark fails when using the ADIOS1 backend (`openPMD::unsupported_data_error`: Unsupported attribute datatype)
- [ ] Feature requests?
- [x] Resolve strange MPI behaviour (see MPIBenchmarkReport.hpp → addReport() method)
- [x] Adjust documentation to refactoring
- [x] Add user documentation (`*.rst`).
- [x] Add licensing information

franzpoeschel committed:
Implement parts of the benchmark configuration

Implementation for slicing a hypercube along one dimension

Refactor MPI benchmarking

Dataset filling is now implemented by an own class instead of a virtual
function.

Add the fundaments for a write benchmark

Adapted from https://openpmd-api.readthedocs.io/en/latest/usage/parallel.html#writing

Nearly complete, but untested MPI benchmark

Missing the collection of benchmarking results for the single threads

Basically executable version of the benchmark

Better handling of dataset indexing

Do not set a compression when the compression string is empty

Fix index calculation

Create the write data anew for every iteration

This allows for bigger benchmarks since only final memory is needed, but
the time for data creation needs to be measured and deducted in the end.
Also document the code.

Refactor the creation of a DataSet

Forward constructor arguments to the DatasetFiller

Add some documentation and cleanup

Flush the Series after every iteration in the benchmark

Integrate with the build system

Allow compiling without MPI enabled

Also fix some shadowed variables

Refactor the benchmark

Make BlockSlicer and DatasetFiller abstract classes instead of
templates.
Add configuration options for backends and numbers of threads to use.
Make it possible to actually include the benchmark.

Refactor benchmark to allow for several datatypes in one run

Make the benchmark system more flexible

Includes setting the number of iterations per benchmark configuration
and adding an adaptor class template SimpleDatasetFillerProvider lifting
a DatasetFiller to a DatasetFillerProvider.

Add documentation for the benchmark

Includes an example usage.

ax3l commented:
Just as a general feature request: it would be cool if this tool allows easy, user-level performance benchmarks on clusters similar to what we did in Figure 1 in https://arxiv.org/abs/1706.00522 . So that we can easily switch backends, compression, written size per node and the like.

franzpoeschel commented:
Thanks for your suggestions, the figure gives a good guideline.
I am not yet fully done (currently having a little fight with MPI), but next to some refactoring I have added support for picking multiple backends and multiple numbers of threads to be used for the benchmark.
The total extent of the written dataset is still fixed during one benchmark to give comparable results. Do you think it would be useful being able to instruct the Benchmark to be executed for several dataset extents?
So in total, the user currently picks
1) compression schemes
2) backends (per filename extension)
3) number of threads to be used

and the benchmark is then executed for every possible combination of entered values.

↑

edit: this has changed now, look into the [documentation](https://github.com/franzpoeschel/openPMD-api/tree/topic-mpiBenchmark/docs/source/utilities)

ax3l commented:
> Do you think it would be useful being able to instruct the Benchmark to be executed for several dataset extents?

Yep, sounds useful since it controls the data per node.

We should write both particle and mesh data, but I am ok to start with the latter.

franzpoeschel commented:
> Yep, sounds useful since it controls the data per node.

I have expressed myself unclearly, the Benchmark class already allows choosing the part of the dataset to write per rank relatively freely. The user can choose this be passing to the constructor an implementation of the [BlockSlicer](https://github.com/openPMD/openPMD-api/blob/0022fbb1709c6088254e1894de6c36785f6d4d87/include/openPMD/benchmark/mpi/BlockSlicer.hpp) abstract class. Once the BlockSlicer is chosen, it is currently fixed to allow for comparable results – unlike compression strategies, backends and thread sizes which can be altered. 

>We should write both particle and mesh data, but I am ok to start with the latter.

To be fair, I just copied that part from [here](https://openpmd-api.readthedocs.io/en/latest/usage/parallel.html), I will look into it.

franzpoeschel committed:
Fix error, resolve MPI misbehaviour

"resolve MPI misbehaviour" = "realized that I had a faulty installation"

ax3l commented on docs/source/index.rst:
```diff
@@ -82,3 +82,13 @@ Development
+   :maxdepth: 1
+   :hidden:
+
+   /utilities/benchmark.rst
```
does not need an initial `/`

ax3l commented on src/Datatype.cpp:
```diff
@@ -144,3 +146,161 @@ std::operator<<(std::ostream& os, openPMD::Datatype d)
+        os << dt;
+        return buf.str();
+    }
+}
```
missing single newline before EOF

ax3l commented on src/benchmark/mpi/OneDimensionalBlockSlicer.cpp:
```diff
@@ -0,0 +1,88 @@
+
+#endif
+}
+
```
one newline too much ;)

ax3l commented on include/openPMD/benchmark/mpi/RandomDatasetFiller.hpp:
```diff
@@ -0,0 +1,129 @@
+
+    };
+
+}
```
missing newline before EOF

ax3l commented on include/openPMD/benchmark/mpi/OneDimensionalBlockSlicer.hpp:
```diff
@@ -0,0 +1,32 @@
+    };
+}
+#endif
+
```
one newline too much ;)

ax3l commented on include/openPMD/benchmark/mpi/MPIBenchmarkReport.hpp:
```diff
@@ -0,0 +1,354 @@
+}
+
+#endif
+
```
one newline too much ;)

ax3l commented on CMakeLists.txt:
```diff
@@ -248,7 +248,8 @@ set(CORE_SOURCE
         src/backend/PatchRecordComponent.cpp
-        src/backend/Writable.cpp)
+        src/backend/Writable.cpp
+        src/benchmark/mpi/OneDimensionalBlockSlicer.cpp)
```
Is it possible that this file and `mpi_benchmark.cpp` should be just in an extra or example target list?

franzpoeschel replied:
I added `mpi_benchmark.cpp` (now `8_mpi_benchmark.cpp`) to the example target list.
For `OneDimensionalBlockSlicer.cpp` do you mean to create a new list for that?

ax3l replied:
Yes, this `.cpp` file is not part of the library's `CORE_SOURCE` but of the examples/apps only.

ax3l commented on docs/source/utilities/mpi_benchmark.cpp:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    MPI_Finalize();
+}
```
missing newline before EOF :)

franzpoeschel committed:
Cleanup

Add licensing information.
Fix whitespaces.
Add the MPI Benchmark example from the documentations to the examples to
be built during openMPI building.

franzpoeschel committed:
Document the example usage better

ax3l commented on include/openPMD/Datatype.hpp:
```diff
@@ -518,6 +519,184 @@ isSame( openPMD::Datatype const d, openPMD::Datatype const e )
+ * openPMD type.
+ */
+
+#if _MSC_VER && !__INTEL_COMPILER
```
Please place this macro block *above* the doxygen block (separated by an empty newline)

ax3l commented:
I can merge this for now, thanks a lot!

ax3l commented:
@franzpoeschel in CI, we get sporadically the error
```
what():  Internal error: Failed to create directories during ADIOS file creation
```

in the MPI benchmark. Can you confirm and might there be a quirk we need to fix?

franzpoeschel commented:
I have seen this error already, yes. I thought it was an issue with the ADIOS backend, I will try finding the cause of it.

franzpoeschel commented:
Just to be sure, #418 is exactly this issue, right?

ax3l commented:
yes, #418 and #421 were sporadic CI issues with parallel I/O :)
